android-project/app/src/main/java/org/diasurgical/devilutionx/DevilutionXSDLActivity.java

```diff
@@ -0,0 +1,136 @@
+import android.content.DialogInterface;
+import android.content.SharedPreferences;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
+import android.preference.PreferenceManager;
+
+import org.libsdl.app.SDLActivity;
+
+import java.io.File;
+
+public class DevilutionXSDLActivity extends SDLActivity {
+	final String permission = Manifest.permission.READ_EXTERNAL_STORAGE;
+	private Handler permissionHandler;
+	private AlertDialog PermissionDialog;
+	private final String SharewareTitle = "Shareware version";
+	private SharedPreferences mPrefs;
+	final private String shouldShowInstallationGuidePref = "shouldShowInstallationGuide";
+
+	protected void onCreate(Bundle savedInstanceState) {
+		checkStoragePermission();
+
+		super.onCreate(savedInstanceState);
+	}
+
+	private boolean shouldShowInstallationGuide(boolean isDataAccessible) {
+		mPrefs = PreferenceManager.getDefaultSharedPreferences(this);
+		return mPrefs.getBoolean(shouldShowInstallationGuidePref, !isDataAccessible);
+	}
+
+	private void installationGuideShown() {
+		SharedPreferences.Editor editor = mPrefs.edit();
+		editor.putBoolean(shouldShowInstallationGuidePref, false);
+		editor.apply();
+	}
+
+	private void checkStoragePermission() {
+		if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP_MR1) {
+			checkStoragePermissionLollipop();
+			return;
+		}
+
+		if (PackageManager.PERMISSION_GRANTED != checkSelfPermission(permission)) {
+			requestPermissions(new String[]{permission}, 1);
+		}
+
+		boolean dataFileFound = hasDataFile();
+		boolean shouldShowInstallationGuide = shouldShowInstallationGuide(dataFileFound && PackageManager.PERMISSION_GRANTED == checkSelfPermission(permission));
+
+		if (!shouldShowInstallationGuide && !shouldShowRequestPermissionRationale(permission)) {
+			return;
+		}
+
+		String message = "To play the full version you must grant the application read access to " + System.getenv("EXTERNAL_STORAGE") + ".";
+		if (!dataFileFound) {
+			message = "To play the full version you must place DIABDAT.MPQ from the original game in " + System.getenv("EXTERNAL_STORAGE") + " and grant the application read access.\n\nIf you don't have the original game then you can buy Diablo from GOG.com.";
+			installationGuideShown();
+		}
+
+		permissionHandler = new Handler();
+
+		PermissionDialog = new AlertDialog.Builder(this)
+				.setTitle(SharewareTitle)
+				.setPositiveButton("OK", new DialogInterface.OnClickListener() {
+					public void onClick(DialogInterface dialog, int id) {
+						permissionHandler.post(new Runnable() {
+							@Override
+							public void run() {
+								throw new RuntimeException();
+							}
+						});
+					}
+				})
+				.setMessage(message)
+				.show();
+
+		if (dataFileFound) {
+			triggerPulse(); // Start full game as soon as we have file access
+		}
+
+		try {
+			Looper.loop();
+		} catch (RuntimeException ignored) {
```

Comment:
I actually hate this. I'm sure there is a better way to exit the looper.

Reply:
Yeah I hate it as well, but wasn't able to find anything better :(

The thing it's trying to do is pause the application until the user closes the dialog. The proper solution is to make custom gui/configure stage and not use the alert dialog, but I also feel strongly that something like that should get out of the way asap and preferably never show if the user correctly added the MPQ files before even launching the app the first time.